File: src/git_commit_ai/backends/ollama.py
"""Ollama backend implementation."""

import httpx

from .base import Backend


class OllamaBackend(Backend):
    """Backend for generating commit messages using Ollama."""

    def __init__(
        self,
        model: str = "llama3.1:8b",
        base_url: str = "http://localhost:11434",
    ):
        self.model = model
        self.base_url = base_url

    def generate(self, prompt: str, temperature: float = 0.7) -> str:
        """Generate a commit message using Ollama API."""
        response = httpx.post(
            f"{self.base_url}/api/generate",
            json={
                "model": self.model,
                "prompt": prompt,
                "temperature": temperature,
                "stream": False,
            },
            timeout=60.0,
        )
        response.raise_for_status()

        data = response.json()
        return data.get("response", "")

    def is_available(self) -> bool:
        """Check if Ollama is running."""
        try:
            response = httpx.get(f"{self.base_url}/api/tags", timeout=5.0)
            return response.status_code == 200
        except httpx.RequestError:
            return False

    def has_model(self, model: str | None = None) -> bool:
        """Check if the specified model is available.

        Args:
            model: Model name to check. Uses self.model if None.

        Returns:
            True if the model is available.
        """
        model = model or self.model
        try:
            response = httpx.get(f"{self.base_url}/api/tags", timeout=5.0)
            if response.status_code != 200:
                return False
            data = response.json()
            models = [m.get("name", "") for m in data.get("models", [])]
            return any(model in m for m in models)
        except httpx.RequestError:
            return False

File: src/backends/ollama.ts
import type { Backend } from "../types.js";

export class OllamaBackend implements Backend {
  private model: string;
  private baseUrl: string;

  constructor(model = "llama3.1:8b", baseUrl = "http://localhost:11434") {
    this.model = model;
    this.baseUrl = baseUrl;
  }

  async generate(prompt: string, temperature = 0.7): Promise<string> {
    const response = await fetch(`${this.baseUrl}/api/generate`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        model: this.model,
        prompt,
        temperature,
        stream: false,
      }),
    });

    if (!response.ok) {
      throw new Error(`Ollama API error: ${response.status}`);
    }

    const data = (await response.json()) as { response?: string };
    return data.response ?? "";
  }

  async isAvailable(): Promise<boolean> {
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), 5000);

      const response = await fetch(`${this.baseUrl}/api/tags`, {
        signal: controller.signal,
      });

      clearTimeout(timeoutId);
      return response.status === 200;
    } catch {
      return false;
    }
  }

  async hasModel(model?: string): Promise<boolean> {
    const checkModel = model ?? this.model;
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), 5000);

      const response = await fetch(`${this.baseUrl}/api/tags`, {
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (response.status !== 200) {
        return false;
      }

      const data = (await response.json()) as {
        models?: Array<{ name?: string }>;
      };
      const models = data.models?.map((m) => m.name ?? "") ?? [];
      return models.some((m) => m.includes(checkModel));
    } catch {
      return false;
    }
  }
}
